### PR TITLE
[X86][AVX512] Use comx for compare

### PR DESCRIPTION
**Pull request by mahesh-attarde**
We added AVX10.2 COMEF ISA in LLVM, This does not optimize correctly in scenario mentioned below.
Summary
Input 
```
define i1 @oeq(float %x, float %y) {
    %1 = fcmp oeq float %x, %y
    ret i1 %1
}define i1 @une(float %x, float %y) {
    %1 = fcmp une float %x, %y
    ret i1 %1
}define i1 @ogt(float %x, float %y) {
    %1 = fcmp ogt float %x, %y
    ret i1 %1
}
// Prior AVX10.2, default code generation

oeq:                                    # @oeq
        cmpeqss xmm0, xmm1
        movd    eax, xmm0
        and     eax, 1
        ret
une:                                    # @une
        cmpneqss        xmm0, xmm1
        movd    eax, xmm0
        and     eax, 1
        ret
ogt:                                    # @ogt
        ucomiss xmm0, xmm1
        seta    al
        ret 
```

This patch will remove `cmpeqss`  and `cmpneqss`. For complete transform check unit test.

Continuing on what PR https://github.com/llvm/llvm-project/pull/113098 added

Earlier Legalization and combine expanded `setcc oeq:ch` node into `and` and `setcc eq` , `setcc o`. From suggestions in community
new internal transform
```
Optimized type-legalized selection DAG: %bb.0 'hoeq:'
SelectionDAG has 11 nodes:
  t0: ch,glue = EntryToken
      t2: f16,ch = CopyFromReg t0, Register:f16 %0
      t4: f16,ch = CopyFromReg t0, Register:f16 %1
    t14: i8 = setcc t2, t4, setoeq:ch
  t10: ch,glue = CopyToReg t0, Register:i8 $al, t14
  t11: ch = X86ISD::RET_GLUE t10, TargetConstant:i32<0>, Register:i8 $al, t10:1

Optimized legalized selection DAG: %bb.0 'hoeq:'
SelectionDAG has 12 nodes:
  t0: ch,glue = EntryToken
        t2: f16,ch = CopyFromReg t0, Register:f16 %0
        t4: f16,ch = CopyFromReg t0, Register:f16 %1
      t15: i32 = X86ISD::UCOMX t2, t4
    t17: i8 = X86ISD::SETCC TargetConstant:i8<4>, t15
  t10: ch,glue = CopyToReg t0, Register:i8 $al, t17
  t11: ch = X86ISD::RET_GLUE t10, TargetConstant:i32<0>, Register:i8 $al, t10:1
```
Earlier transform is mentioned here https://github.com/llvm/llvm-project/pull/113098#discussion_r1810307663

### CODE REVIEW

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 1]
```diff
@@ -24073,6 +24078,13 @@ SDValue X86TargetLowering::LowerSETCC(SDValue Op, SelectionDAG &DAG) const {
     return IsStrict ? DAG.getMergeValues({Res, Chain}, dl) : Res;
   }
 
+  if (Subtarget.hasAVX10_2_512()) {
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:24081]: `hasAVX10_2`

File: llvm/test/CodeGen/X86/avx10_2-cmp.ll [hunk 2]
```diff
@@ -0,0 +1,237 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc < %s -mtriple=x86_64-unknown-unknown -mattr=+avx10.2-512 | FileCheck %s --check-prefix=X64
+; RUN: llc < %s -mtriple=i386-unknown-unknown -mattr=+avx10.2-512 | FileCheck %s --check-prefix=X86
```
Comment (phoebewang) [inline llvm/test/CodeGen/X86/avx10_2-cmp.ll:3]: `+avx10.2`

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 3]
```diff
@@ -218,10 +218,14 @@ X86TargetLowering::X86TargetLowering(const X86TargetMachine &TM,
...
-    setCondCodeAction(ISD::SETOEQ, VT, Expand);
-    setCondCodeAction(ISD::SETUNE, VT, Expand);
+  for (auto VT : {MVT::f32, MVT::f64}) {
+    setCondCodeAction(ISD::SETOEQ, VT,
+                      Subtarget.hasAVX10_2() ? Custom : Expand);
+    setCondCodeAction(ISD::SETUNE, VT,
+                      Subtarget.hasAVX10_2() ? Custom : Expand);
   }
+  setCondCodeAction(ISD::SETOEQ, MVT::f80, Expand);
+  setCondCodeAction(ISD::SETUNE, MVT::f80, Expand);
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:228]: You can leave it as is and override later with
```
if (Subtarget.hasAVX10_2()) {
  for (auto VT : {MVT::f16, MVT::f32, MVT::f64}) {
    ...
}
```

File: llvm/lib/Target/X86/X86InstrAVX10.td [hunk 4]
```diff
@@ -1564,6 +1582,16 @@ multiclass avx10_com_ef_int<bits<8> Opc, X86VectorVTInfo _, SDNode OpNode,
 }
 
 let Defs = [EFLAGS], Uses = [MXCSR], Predicates = [HasAVX10_2] in {
+
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86InstrAVX10.td:1585]: Remove blank line.

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 5]
```diff
@@ -2292,8 +2298,10 @@ X86TargetLowering::X86TargetLowering(const X86TargetMachine &TM,
     setOperationAction(ISD::FP_EXTEND,            MVT::f32, Legal);
     setOperationAction(ISD::STRICT_FP_EXTEND,     MVT::f32, Legal);
 
-    setCondCodeAction(ISD::SETOEQ, MVT::f16, Expand);
-    setCondCodeAction(ISD::SETUNE, MVT::f16, Expand);
+    setCondCodeAction(ISD::SETOEQ, MVT::f16,
+                      Subtarget.hasAVX10_2() ? Custom : Expand);
+    setCondCodeAction(ISD::SETUNE, MVT::f16,
+                      Subtarget.hasAVX10_2() ? Custom : Expand);
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:2304]: Why don't combine them to the loop? You can move them to line 2445.

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 6]
```diff
@@ -2442,7 +2450,6 @@ X86TargetLowering::X86TargetLowering(const X86TargetMachine &TM,
       }
     }
   }
-
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:2445]: Unrelated change.

**Review by phoebewang — APPROVED** [review]
LGTM with 2 nits.

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 5 above)
Comment (mahesh-attarde) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:2304]: I can do that. Although `setGroup(MVT::f16);` on 2286 line above has better proximity fore related code.

Comment (phoebewang) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:2304]: The code is in the scope of `hasFP16`. We prefer to organizing them in feature bulk rather than mixing them together.

Comment (mahesh-attarde) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:2304]: This code movement result in 3 regressions.

**Comment by mahesh-attarde** [comment]
Can we merge this?

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 5 above)
Comment (phoebewang) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:2304]: I don't see how it results in regressions. Did you remove these 2 Expand?

Comment (mahesh-attarde) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:2304]: Here is link on infra.
https://buildkite.com/llvm-project/github-pull-requests/builds/113138#0192c4a7-5933-4d41-b189-f83fdfe10fd9

Comment (phoebewang) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:2304]: I meant this:
```
diff --git a/llvm/lib/Target/X86/X86ISelLowering.cpp b/llvm/lib/Target/X86/X86ISelLowering.cpp
index 676ae6b87610..1ffec84d9958 100644
--- a/llvm/lib/Target/X86/X86ISelLowering.cpp
+++ b/llvm/lib/Target/X86/X86ISelLowering.cpp
@@ -2441,6 +2441,11 @@ X86TargetLowering::X86TargetLowering(const X86TargetMachine &TM,
         setOperationAction(ISD::SETCC, VT, Custom);
       }
     }
+
+    for (auto VT : {MVT::f16, MVT::f32, MVT::f64}) {
+      setCondCodeAction(ISD::SETOEQ, VT, Custom);
+      setCondCodeAction(ISD::SETUNE, VT, Custom);
+    }
   }

   if (!Subtarget.useSoftFloat() && Subtarget.hasVLX()) {
```

Notice it's under these code, so it should not affect any existing tests.

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 7]
```diff
@@ -223,6 +223,12 @@ X86TargetLowering::X86TargetLowering(const X86TargetMachine &TM,
     setCondCodeAction(ISD::SETUNE, VT, Expand);
   }
 
+  if (Subtarget.hasAVX10_2()) {
+    for (auto VT : {MVT::f16, MVT::f32, MVT::f64}) {
+      setCondCodeAction(ISD::SETOEQ, VT, Custom);
+      setCondCodeAction(ISD::SETUNE, VT, Custom);
+    }
+  }
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:231]: Move this under FP16 code. We need to override `f16` with this code.

Comment (mahesh-attarde) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:231]: you mean only FP16 or Complete SetCondCodeAction for f16, f32 and f64?

Comment (phoebewang) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:231]: All of them. See the line number of previous diff.

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 5 above)
Comment (mahesh-attarde) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:2304]: https://github.com/llvm/llvm-project/pull/113695
https://github.com/llvm/llvm-project/pull/113905,  2 commits on this very branch
I tried out  4 different ways, except current change i get regression for all moves.

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 7 above)
Comment (mahesh-attarde) [reply llvm/lib/Target/X86/X86ISelLowering.cpp]: https://github.com/llvm/llvm-project/pull/113905 I tried this. Moved SETOEQ and SETUNE group after avx10_2 & softfloat feature. it did not work.

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 5 above)
Comment (phoebewang) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:2304]: See https://github.com/llvm/llvm-project/commit/2c9b981b9d8541d7c9e026baff1325e1708caeae
I don't see any problem on my side.
Maybe you was misled. The diff above was to demonstrate the change in function `X86TargetLowering`, so I omitted the change in `LowerSETCC`. I'm surprised you didn't explore along this way.

**Comment by phoebewang** [comment]
This revision LG. @mahesh-attarde can you update the description before we merging it?

**Comment by mahesh-attarde** [comment]
> This revision LG. @mahesh-attarde can you update the description before we merging it?

done.
